lib/openssh/sshd.go

```diff
@@ -0,0 +1,210 @@
+		return trace.ConvertSystemError(err)
+	}
+
+	if restart {
+		b.sshd.restart()
+	}
+
+	return nil
+}
+
+func WriteKeys(keysdir string, id *auth.Identity, cas []types.CertAuthority) error {
+	if err := os.WriteFile(filepath.Join(keysdir, TeleportKey), id.KeyBytes, 0600); err != nil {
+		return trace.ConvertSystemError(err)
+	}
+	if err := os.WriteFile(filepath.Join(keysdir, TeleportCert), id.CertBytes, 0600); err != nil {
+		return trace.ConvertSystemError(err)
+	}
+
+	var caKeyBytes []byte
+	for _, ca := range cas {
+		for _, key := range ca.GetTrustedSSHKeyPairs() {
+			caKeyBytes = append(caKeyBytes, key.PublicKey...)
+		}
+	}
+
+	if err := os.WriteFile(filepath.Join(keysdir, TeleportOpenSSHCA), caKeyBytes, 0600); err != nil {
+		return trace.ConvertSystemError(err)
+	}
+	return nil
+}
+
+func TryCreateDefaultAgentlesKeysDir(agentlessKeysPath string) error {
+	_, err := os.Stat(agentlessKeysPath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return trace.Wrap(os.MkdirAll(agentlessKeysPath, 0700))
+		}
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+type sshdBackend struct{}
+
+var _ sshdBackendOperations = &sshdBackend{}
+
+func (*sshdBackend) checkConfig(path string) error {
+	cmd := exec.Command(sshdBinary, "-t", "-f", path)
+	if err := cmd.Run(); err != nil {
+		return trace.Wrap(err, "teleport generated an invalid ssh config file, not writing")
+	}
+	return nil
+}
+
+func (*sshdBackend) restart() error {
+	cmd := exec.Command("sshd", "-t")
+	if err := cmd.Run(); err != nil {
+		return trace.Wrap(err, "teleport generated an invalid ssh config file")
+	}
+
+	cmd = exec.Command("systemctl", "restart", "sshd")
+	if err := cmd.Run(); err != nil {
+		return trace.Wrap(err, "teleport failed to restart the sshd service")
+	}
+	return nil
+}
+
+type SSHdConfigUpdate struct {
+	SSHdConfigPath         string
+	TeleportSSHDConfigPath string
+	OpenSSHCAPath          string
+
+	HostKeyPaths  []string
+	HostCertPaths []string
+}
+
+func prefixSSHDConfig(sshdConfigPath, config string) error {
+	contents, err := os.ReadFile(sshdConfigPath)
+	if err != nil {
+		return trace.ConvertSystemError(err)
+	}
+	line := append([]byte(config), byte('\n'))
+	contents = append(line, contents...)
+
+	err = os.WriteFile(sshdConfigPath, contents, 0644)
+	if err != nil {
+		return trace.ConvertSystemError(err)
+	}
+
+	return nil
+}
+
+func checkSSHDConfigAlreadyUpdated(sshdConfigPath, fileContains string) (bool, error) {
+	contents, err := os.ReadFile(sshdConfigPath)
+	if err != nil {
+		return false, trace.ConvertSystemError(err)
+	}
+	return !strings.Contains(string(contents), fileContains), nil
+}
+
+func fmtSSHDConfigUpdate(openSSHCAPath string, hostKeyPaths, hostCertPaths []string) string {
+	return fmt.Sprintf(`%s
+TrustedUserCaKeys %s
+HostKey %s
+HostCertificate %s
+`,
+		sshdConfigSectionModificationHeader,
+		openSSHCAPath,
+		strings.Join(hostKeyPaths, " "),
+		strings.Join(hostCertPaths, " "),
```

Comment:
I don't think `HostKey` takes multiple paths like that; even if it did, `sshd` will only ever present a single key per signature algo, I'm afraid, so multiple `ssh-cert-rsa-v01@openssh.com` certs are going to be useless. 😢